File: accounts-service/models/tarjetas_plastico.py
from __future__ import annotations

from datetime import date

from sqlalchemy import Date, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from utils.database import Base


class TarjetasPlastico(Base):
    __tablename__ = "tarjetas_plastico"

    id_tarjeta: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    id_cuenta: Mapped[int] = mapped_column(ForeignKey("cuenta_bancaria.id_cuenta"), nullable=False)
    numero_tarjeta: Mapped[str] = mapped_column(String(16), nullable=False)
    mes_expira: Mapped[int] = mapped_column(Integer, nullable=False)
    anio_expira: Mapped[int] = mapped_column(Integer, nullable=False)
    cvc: Mapped[str] = mapped_column(String(3), nullable=False)
    fecha_emision: Mapped[date] = mapped_column(Date, nullable=False)
    estatus: Mapped[str] = mapped_column(String(20), nullable=False, default="activa")
    nip: Mapped[str | None] = mapped_column(String(4), nullable=True)

    cuenta_bancaria: Mapped["CuentaBancaria"] = relationship("CuentaBancaria", back_populates="tarjetas_plastico")

File: auth-service/src/main/java/banksoft/model/TarjetasPlastico.java
package banksoft.model;

import jakarta.persistence.*;
import java.time.LocalDate;

@Entity
@Table(name = "tarjetas_plastico")
public class TarjetasPlastico {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id_tarjeta")
    private Integer idTarjeta;

    @Column(name = "id_cuenta", nullable = false)
    private Integer idCuenta;

    @Column(name = "numero_tarjeta", nullable = false, length = 16)
    private String numeroTarjeta;

    @Column(name = "mes_expira", nullable = false)
    private Integer mesExpira;

    @Column(name = "anio_expira", nullable = false)
    private Integer anioExpira;

    @Column(name = "cvc", nullable = false, length = 3)
    private String cvc;

    @Column(name = "fecha_emision", nullable = false)
    private LocalDate fechaEmision;

    @Column(name = "estatus", nullable = false, length = 20)
    private String estatus = "activa";

    @Column(name = "nip", length = 4)
    private String nip;
 

    public TarjetasPlastico() {}

    public TarjetasPlastico(Integer idCuenta, String numeroTarjeta, Integer mesExpira, Integer anioExpira, String cvc) {
        this.idCuenta = idCuenta;
        this.numeroTarjeta = numeroTarjeta;
        this.mesExpira = mesExpira;
        this.anioExpira = anioExpira;
        this.cvc = cvc;
        this.fechaEmision = LocalDate.now();
        this.estatus = "activa";
    }

    public Integer getIdTarjeta() { return idTarjeta; }
    public void setIdTarjeta(Integer idTarjeta) { this.idTarjeta = idTarjeta; }

    public Integer getIdCuenta() { return idCuenta; }
    public void setIdCuenta(Integer idCuenta) { this.idCuenta = idCuenta; }

    public String getNumeroTarjeta() { return numeroTarjeta; }
    public void setNumeroTarjeta(String numeroTarjeta) { this.numeroTarjeta = numeroTarjeta; }

    public Integer getMesExpira() { return mesExpira; }
    public void setMesExpira(Integer mesExpira) { this.mesExpira = mesExpira; }

    public Integer getAnioExpira() { return anioExpira; }
    public void setAnioExpira(Integer anioExpira) { this.anioExpira = anioExpira; }

    public String getCvc() { return cvc; }
    public void setCvc(String cvc) { this.cvc = cvc; }

    public LocalDate getFechaEmision() { return fechaEmision; }
    public void setFechaEmision(LocalDate fechaEmision) { this.fechaEmision = fechaEmision; }

    public String getEstatus() { return estatus; }
    public void setEstatus(String estatus) { this.estatus = estatus; }

    public String getNip() { return nip; }
    public void setNip(String nip) { this.nip = nip; }
 
}
